### PR TITLE
Genetic search algorithms

### PR DESCRIPTION
**Pull request by nicolaefilat**

### CODE REVIEW

**Comment by nicolaefilat** [comment]
I messed up with git here I wanted to delete some commit and no new commits were left and that is why it was automatically closed.

File: src/genetic_functions/crossover.jl [hunk 1]
```diff
@@ -0,0 +1,48 @@
+function crossover_2_children(parent1::RuleNode, parent2::RuleNode)
```
Comment (sebdumancic) [inline src/genetic_functions/crossover.jl:1]: lets name these more informatively. I like the format of keeping the role of the function within GA as the first part of the name: `crossover_X`, `mutate_X`, `select_X`.
So maybe we call this one `crossover_swap_children_2`?

Comment (nicolaefilat) [reply src/genetic_functions/crossover.jl:1]: Yes I will rename the function name. But I am not sure if `crossover_swap_children_2` indicates that two children are returned.

File: src/genetic_functions/mutation.jl [hunk 2]
```diff
@@ -0,0 +1,6 @@
+function random_mutate!(program::RuleNode,grammar::Grammar, max_depth::Int64 = 2)
```
Comment (sebdumancic) [inline src/genetic_functions/mutation.jl]: here as well -> `mutate_random!`

File: src/genetic_search_iterator.jl [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+    max_depth::Int64  = 10    # not used
+end
+
+struct GeneticIteratorState
+    population::Array{RuleNode}
+end
+
+Base.IteratorSize(::GeneticSearchIterator) = Base.SizeUnknown()
+Base.eltype(::GeneticSearchIterator) = RuleNode
```
Comment (sebdumancic) [inline src/genetic_search_iterator.jl:34]: is this `RuleNode` or `Array{RuleNode}`?

Comment (nicolaefilat) [reply src/genetic_search_iterator.jl:34]: Right now is just `RuleNode`.

File: src/genetic_functions/mutation.jl
Comment (THinnerichs) [inline src/genetic_functions/mutation.jl:1]: Please add typecheck. For now the new random program could be inserted at an infeasible location. Easy fix: get the type of node_location and generate the program of that type (rand has that functionality already).

File: src/genetic_functions/select_parents.jl [hunk 4]
```diff
@@ -0,0 +1,28 @@
+function select_fitness_proportional_parents(population::Array{RuleNode}, fitness_array::Array{<:Real})::Tuple{RuleNode,RuleNode}
+    sum_of_fitness = sum(fitness_array)
+    fitness_array_normalized = [fitness_value / sum_of_fitness for fitness_value in fitness_array]
+    parent1 = select_chromosome(population, fitness_array_normalized)
+    parent2 = select_chromosome(population, fitness_array_normalized)
+    return parent1, parent2
+end
+
+function select_two_random_parents(population::Array{RuleNode}, fitness_array::Array{<:Real})::Tuple{RuleNode,RuleNode}
```
Comment (THinnerichs) [inline src/genetic_functions/select_parents.jl:9]: What is the difference between `select_fitness_proportional_parents` and `select_two_random_parents`?

File: src/genetic_search_iterator.jl [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        end
+    end
+
+    # return the program that has the highest fitness
+    return (new_population[begin], GeneticIteratorState(new_population))
+end
+
+
+function calculate_cost_array(program::RuleNode, examples::Vector{<:Example}, grammar::Grammar, evaluation_function::Function)
+    results = Tuple{<:Number,<:Number}[]
```
Comment (THinnerichs) [inline src/genetic_search_iterator.jl:142]: Should results be just numbers? That should be the same type as in the `HerbData.IOExample` as we potentially want to use this algorithm on arbitrary examples.

File: src/genetic_functions/select_parents.jl (hunk 4 above)
Comment (THinnerichs) [reply src/genetic_functions/select_parents.jl:9]: I removed `select_two_random_parents` for now.

**Review by THinnerichs — APPROVED** [review]

**Pull request merged**